### PR TITLE
[SGD] add Horovod backend

### PR DESCRIPTION
**Pull request by matthewdeng**
## Why are these changes needed?
Implements `Horovod` to use as a backend for distributed training. This is heavily influenced by the existing [Horovod on Ray](https://horovod.readthedocs.io/en/stable/ray_include.html) implementation and leverages some components.

## Implementation Details
1. Implement logic from [_ExecutorDriver.start](https://github.com/horovod/horovod/blob/master/horovod/ray/runner.py#L358) in `HorovodBackend.on_start`. **Minor Note:** `extra_env_vars` is removed as we have `initialization_hook`.
2. Leverage Horovod on Ray's [Coordinator](https://github.com/horovod/horovod/blob/master/horovod/ray/runner.py#L41) for setting up rendezvous.
3. Move Horovod specific logic implemented in [BaseHorovodWorker](https://github.com/horovod/horovod/blob/master/horovod/ray/worker.py) into the `HorovodBackend` to keep `BaseWorker` generic.
4. `HorovodConfig` replaces [MiniSettings](https://github.com/horovod/horovod/blob/master/horovod/ray/runner.py#L18).
5. Added a `horovod_example` modified from the existing `python/ray/util/horovod/horovod_example.py` - made some fixes to get the original example to run as well.
6. Added a `BackendExecutor._setup_gpus` method which sets the appropriate `CUDA_VISIBLE_DEVICES` environment variable on each worker, so that workers on the same node can communicate with each other.

## Future steps
The following features will be added to RaySGD while maintaining compatibility with Horovod on Ray's existing functionality.
1. **Placement Groups:** [Strategies](https://github.com/horovod/horovod/blob/master/horovod/ray/strategy.py)
2. **Class API:**  [Stateful Execution](https://horovod.readthedocs.io/en/stable/ray_include.html#stateful-execution)
3. **Elasticity:** [Elastic Ray Executor](https://horovod.readthedocs.io/en/stable/ray_include.html#elastic-ray-executor)

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/util/horovod/horovod_example.py [hunk 1]
```diff
@@ -118,7 +118,8 @@ def train_fn(data_dir=None,
 def main(num_workers, use_gpu, **kwargs):
     settings = RayExecutor.create_settings(timeout_s=30)
     executor = RayExecutor(settings, use_gpu=use_gpu, num_workers=num_workers)
-    executor.run(train_fn, kwargs=kwargs)
+    executor.start()
```
Comment (amogkam) [inline python/ray/util/horovod/horovod_example.py:121]: Nice catch!

File: python/ray/util/sgd/v2/examples/horovod_example.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+import horovod.torch as hvd
+import ray
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+import torch.utils.data.distributed
+from filelock import FileLock
+from ray.util.sgd.v2 import Trainer
+from torchvision import datasets, transforms
+
```
Comment (amogkam) [inline python/ray/util/sgd/v2/examples/horovod_example.py:13]: Can you add this example to CI?

File: python/ray/util/sgd/v2/tests/test_trainer.py [hunk 3]
```diff
@@ -538,6 +542,30 @@ def test_torch_fashion_mnist_gpu(ray_start_2_cpus_2_gpus):
...
+    num_workers = 2
+    trainer = Trainer("horovod", num_workers)
+    trainer.start()
+    result = trainer.run(simple_fn)
+    trainer.shutdown()
+
+    assert result == list(range(num_workers))
+
+
+def test_horovod_torch_mnist(ray_start_2_cpus):
```
Comment (amogkam) [inline python/ray/util/sgd/v2/tests/test_trainer.py:559]: Can you also add a GPU unit test for Horovod?

File: python/ray/util/sgd/v2/backends/horovod.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+    See https://github.com/horovod/horovod/blob/master/horovod/runner/common/util/settings.py # noqa: E501
+
+    Args:
+        nics (int): Network interfaces that can be used for communication.
+        verbose (int): Horovod logging verbosity.
+    """
+    # timeout_s: int = 300
+    nics: set = None
+    verbose: int = 1
```
Comment (amogkam) [inline python/ray/util/sgd/v2/backends/horovod.py:28]: Are the other settings not required here?

File: python/ray/util/sgd/v2/backends/horovod.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+class HorovodConfig(BackendConfig):
+    """Configurations for Horovod setup.
+
+    See https://github.com/horovod/horovod/blob/master/horovod/runner/common/util/settings.py # noqa: E501
+
+    Args:
+        nics (int): Network interfaces that can be used for communication.
+        verbose (int): Horovod logging verbosity.
+    """
+    # timeout_s: int = 300
```
Comment (amogkam) [inline python/ray/util/sgd/v2/backends/horovod.py:26]: Should this be uncommented?

File: python/ray/util/sgd/v2/backends/horovod.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    return ray.get_runtime_context().node_id.hex()
+
+
+def get_hostname() -> str:
+    # TODO: This is probably not the right way to retrieve
+    # the intended hostname.
+    return socket.gethostname()
+
+
+def update_env_vars(env_vars: Dict[str, str]):
```
Comment (amogkam) [inline python/ray/util/sgd/v2/backends/horovod.py:45]: Same as above, we can probably move this to `WorkerGroup`

File: python/ray/util/sgd/v2/backends/horovod.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    # timeout_s: int = 300
+    nics: set = None
+    verbose: int = 1
+
+    @property
+    def backend_cls(self):
+        return HorovodBackend
+
+
+def get_node_id() -> str:
```
Comment (amogkam) [inline python/ray/util/sgd/v2/backends/horovod.py:35]: This is probably general enough to move to WorkerGroup, right?

**Review by amogkam — COMMENTED** [review]
Nice! I think there's some logic missing for `CUDA_VISIBLE_DEVICES` though.

In Horovod-Ray we manually override the `CUDA_VISIBLE_DEVICES` environment variable on each worker to include the GPU ids for the entire node: https://github.com/horovod/horovod/blob/283549c1c5218ca8919e7f7e897dda6a69c02b7d/horovod/ray/strategy.py#L117-L133.

More context: https://github.com/horovod/horovod/pull/2218#discussion_r481349578

**Comment by richardliaw** [comment]
Actually, we should be setting CUDA visible devices like that for all Trainers:

https://github.com/ray-project/ray/issues/16979

> https://horovod.readthedocs.io/en/stable/troubleshooting_include.html#running-out-of-memory

Setting CUDA_VISIBLE_DEVICES has additional disadvantage for GPU version - CUDA will not be able to use IPC, which will likely cause NCCL and MPI to fail.

**Comment by amogkam** [comment]
Actually yeah, good point. @matthewdeng can we add this logic to `BackendConfig` itself, instead of just `HorovodConfig`?

**Review by amogkam — COMMENTED** [review]
Looks good! Just left a few comments. This works on GPU right? Do you know if the GPU CI is ready for use?

File: python/ray/util/sgd/v2/backends/horovod.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+
+@dataclass
+class HorovodConfig(BackendConfig):
+    """Configurations for Horovod setup.
+
+    See https://github.com/horovod/horovod/blob/master/horovod/runner/common/util/settings.py # noqa: E501
+
+    Args:
+        nics (int): Network interfaces that can be used for communication.
```
Comment (richardliaw) [inline python/ray/util/sgd/v2/backends/horovod.py:22]: nics should be a set of strings right?

**Review by richardliaw — COMMENTED** [review]
Took a brief look, looks good to me, placement group strategies shouldn't belong in the Horovod backend right?

File: .buildkite/pipeline.yml [hunk 9]
```diff
@@ -411,7 +411,7 @@
   conditions: ["RAY_CI_SGD_AFFECTED"]
   commands:
     - cleanup() { if [ "${BUILDKITE_PULL_REQUEST}" = "false" ]; then ./ci/travis/upload_build_info.sh; fi }; trap cleanup EXIT
-    - SGD_TESTING=1 ./ci/travis/install-dependencies.sh
+    - SGD_TESTING=1 INSTALL_HOROVOD=1 ./ci/travis/install-dependencies.sh
```
[the remark this answers is not in the document]
Comment (amogkam) [reply .buildkite/pipeline.yml:414]: Never mind, you can ignore this. Looks the the installation is successful on the 3.6 build

File: python/ray/util/sgd/v2/backends/backend.py [hunk 10]
```diff
@@ -104,8 +106,66 @@ def start(self, initialization_hook: Optional[Callable[[], None]] = None):
...
+            def get_node_id_and_gpu():
+                node_id = get_node_id()
+                gpu_ids = get_gpu_ids()
+                return node_id, gpu_ids
+
+            node_ids_and_gpu_ids = self.worker_group.execute(
+                get_node_id_and_gpu)
+
+            node_id_to_worker_id = defaultdict(list)
+            node_id_to_gpu_ids = defaultdict(list)
```
[the remark this answers is not in the document]
Comment (matthewdeng) [reply python/ray/util/sgd/v2/backends/backend.py]: Good point! Updated this accordingly and added tests.

File: python/ray/util/sgd/v2/tests/test_backend.py [hunk 11]
```diff
@@ -249,6 +264,21 @@ def check_process_group():
     assert not any(e.finish_training())
 
 
+def test_cuda_visible_devices(ray_2_node_2_gpu):
```
[the remark this answers is not in the document]
Comment (matthewdeng) [reply python/ray/util/sgd/v2/tests/test_backend.py]: Added!

File: python/ray/util/sgd/v2/worker_group.py [hunk 12]
```diff
@@ -183,3 +186,28 @@ def execute_single(self, worker_index: int, func: Callable[..., T], *args,
 
     def __len__(self):
         return len(self.workers)
+
```
[the remark this answers is not in the document]
Comment (matthewdeng) [reply python/ray/util/sgd/v2/worker_group.py]: Yeah I kept it out of the class to be consistent with the `execute` pattern. Moved to `utils.py`.

File: python/ray/util/sgd/v2/backends/horovod.py (hunk 8 above)
Comment (matthewdeng) [reply python/ray/util/sgd/v2/backends/horovod.py]: Yep, fixed!

**Comment by matthewdeng** [comment]
@amogkam 
> This works on GPU right? 

Yep! I added a cluster config in this PR that can be used for testing:

```bash
ray up horovod-cluster.yaml
# CPU
ray submit horovod-cluster.yaml horovod_example.py 
# GPU
ray submit horovod-cluster.yaml horovod_example.py --use-gpu
```
> Do you know if the GPU CI is ready for use?

Yeah, I will follow up in a separate PR to port over SGD tests to run on GPU.

@richardliaw 
> placement group strategies shouldn't belong in the Horovod backend right?

Yeah, I just put the TODO in `horovod.py` because it is a "gap" compared to Horovod on Ray implementation, and wanted to explicitly call this out until we support placement groups in SGD.

File: python/ray/util/sgd/v2/backends/horovod.py [hunk 13]
```diff
@@ -0,0 +1,89 @@
+import logging
+import os
+from dataclasses import dataclass
+from typing import Optional, Set
+
+import ray
+from horovod.ray.runner import Coordinator
```
Comment (amogkam) [inline python/ray/util/sgd/v2/backends/horovod.py:7]: Can you make the horovod imports optional, like in `TorchConfig` and `TFConfig`?

**Review by richardliaw — COMMENTED** [review]
This looks good to me; will leave to Amog to approve/mege

**Review by amogkam — APPROVED** [review]
LGTM!

**Pull request merged**